src/dag_algo/mod.rs

```diff
@@ -634,3 +636,50 @@ pub fn collect_bicolor_runs(
 
     Ok(block_list)
 }
+
+/// Return the transitive closure of a directed acyclic graph
+///
+/// The transitive closure of :math:`G = (V, E)` is a graph :math:`G+ = (V, E+)`
+/// such that for all pairs of :math:`v, w` in :math:`V` there is an edge
+/// :math:`(v, w) in :math:`E+` if and only if there is a non-null path from
+/// :math:`v` to :math:`w` in :math:`G`.
+///
+/// :param PyDiGraph graph: The input DAG to compute the transitive closure of
+/// :param list topological_order: An optional topological order for ``graph``
+///     which represents the order the graph will be traversed in computing
+///     the transitive closure. If one is not provided (or it is explicitly
+///     set to ``None``) a topological order will be computed by this function.
+///
+/// :returns: The transitive closure of ``graph``
+/// :rtype: PyDiGraph
+///
+/// :raises DAGHasCycle: If the input ``graph`` is not acyclic
+#[pyfunction]
+#[pyo3(text_signature = "(graph, / topological_order=None)")]
+pub fn transitive_closure_dag(
```

Comment:
Would you mind implementing this as a rustworkx-core function? I think it would make reusing it for #704 easier

Reply:
There is [`Build`](https://docs.rs/petgraph/latest/petgraph/data/trait.Build.html) trait.

Reply:
Thanks @georgios-ts I completely missed the build trait that should do the trick.

I'm working on a local branch right now, but I'm fighting the trait implementations now. Basically `IntoNodeIdentifiers` and the other visit traits are implemented for `&StablePyGraph` but not a mutable borrow or the type itself and `Build` is only available on a mutable borrow or the type itself. I'll see what I can come up with

Reply:
I've faced a similar situation before and it's quite tricky (and ugly) to get this right. I think a signature like this
```rust
pub fn transitive_closure_dag<'a, G, F>(
    graph: &'a G,
    topological_order: Option<Vec<G::NodeId>>,
    default_weight_fn: F,
) -> Result<G, algo::Cycle<G::NodeId>>
where
    G: Clone + Build,
    for<'b> &'b G:
        GraphBase<NodeId = G::NodeId> + Visitable + IntoNeighborsDirected + IntoNodeIdentifiers,
    G::NodeId: std::cmp::Eq + std::hash::Hash,
    F: Fn() -> G::EdgeWeight,
{
...
}
```
will do the job.

Reply:
Awesome, thanks for the hint. That was enough to get it working for me. I didn't realize you could specify lifetimes on generic constraints like that, it was the piece I was missing along with defining `&G: GraphBase<NodeId = G::NodeId>` that was enough to get it moved into rustworkx-core in https://github.com/Qiskit/rustworkx/pull/707/commits/c4c0ab73444c01681402115f2a561b5f320d3f60